### ISSUE TITLE
Select VSCode Server volume by custom name

### ISSUE THREAD

**Issue by kennethredler** [microsoft/vscode-remote-release#5281]
### Feature Request

Please allow the "Remote - Containers" extension workflow to function for multiple users on the same Docker host by facilitating deconflicting the name of the volume mounted on the dev container into which the plugin installs VSCode server components.

#### Potential solutions

1. Use a GUID/UUID in the volume name.
2. Allow the user to customize the name of the volume or volume mount entirely in the same manner as is supported for the workspace volume.

### Back Story

When "[Developing inside a container on a remote Docker host](https://code.visualstudio.com/docs/remote/containers-advanced#_developing-inside-a-container-on-a-remote-docker-host)" such as [Mirantis Container Runtime](https://www.mirantis.com/software/container-runtime/) (formerly Docker Engine - Enterprise) using the first variation 
documented where 

> **create your remote dev container first**, and then **clone your source code into a named volume** since this does not require you to have direct access to the filesystem on the remote host

...a preexisting, custom-named volume into which code is cloned can be selected using [the `workspaceMount` directive in `devcontainer.json`](https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_devcontainerjson-properties):

> ```json
> {
>  ...
>  "workspaceMount": "source=remote-workspace,target=/workspace,type=volume"
> }

However, when the container is run, a second volume is mounted named `vscode`. This is not configurable and if, in solutions like Docker EE, that volume already exists and is in a Docker EE "Collection" that the current user is not a member. It seems the named volumes in Docker EE are global so only one person could use this "Developing inside a container on a remote Docker host" capability _per Docker EE UCP!!!_

Example of failure:

```log
[PID 16260] [9523 ms] Start: Starting container
[PID 16260] [9523 ms] Start: Run: docker run --sig-proxy=false -a STDOUT -a STDERR --mount source=user,target=/workspace,type=volume --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=c:\Users\user\wa -l vsch.quality=stable -l vsch.remote.devPort=0 -e NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/my.internal.crt --entrypoint /bin/sh vsc-wa-97082f941e1c26e9edb32bd1423ad8e7 -c echo Container started ; trap "exit 0" 15; while sleep 1 & wait $!; do :; done
[PID 16260] 
[PID 16260] docker: Error response from daemon: access denied:
no access to Volume Create/Attach, on collection a98b123d-e8c9-2a1a-a11a-2b43dc456134.
See 'docker run --help'.
[PID 16260] [11775 ms] Start: Run: docker ps -q -a --filter label=vsch.local.folder=c:\Users\user\wa --filter label=vsch.quality=stable
[PID 16260] [13596 ms] Command failed: docker run --sig-proxy=false -a STDOUT -a STDERR --mount source=user,target=/workspace,type=volume --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=c:\Users\user\wa -l vsch.quality=stable -l vsch.remote.devPort=0 -e NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/my.internal.crt --entrypoint /bin/sh vsc-wa-97082f941e1c26e9edb32bd1423ad8e7 -c echo Container started ; trap "exit 0" 15; while sleep 2 & wait $!; do :; done
```

| Component | Version |
| ------------ | ------- |
| Windows     | 10 |
| VSCode       | 1.55.2   |
| Remote - Containers extension | 0.166.1 |
| Docker CLI   | 20.10.7 |

**Comment by chrmarti** [comment]
You can turn the use of that volume off:
![image](https://user-images.githubusercontent.com/9205389/124493194-a3b06e80-ddb5-11eb-9a35-fdbdd32a526f.png)

Does that solve the issue?

**Comment by kennethredler** [comment]
Thank you for the suggestion, @chrmarti .

Yes that does workaround the issue as the `vscode` volume is no longer mounted.
You've "unblocked" my workflow, enabling this solution to scale. Thanks!!! 👏 

The remaining value of this feature request would be to allow the user to prescribe the name of the Cache Volume so as to retain the server caching functionality instead of forfeiting it.

**Comment by kennethredler** [comment]
Also, is this "Cache Volume" configuration item documented? [Other relevant documentation](https://code.visualstudio.com/docs/remote/containers-advanced#_avoiding-extension-reinstalls-on-container-rebuild) seems to indicate alternate behavior not mentioning this default caching:

   > By default, VS Code will install extensions and VS Code Server **inside the container's filesystem**. While this has performance benefits over a locally mounted filesystem, the disadvantage is that **VS Code will have to reinstall them on a container rebuild**. If you find yourself rebuilding frequently, you can use a local "named volume" mount so that the extensions and VS Code Server survive a container rebuild.
   > ...
   > 2. Next, we'll configure a named volume mount for `~/.vscode-server/extensions` and `~/.vscode-server-insiders/extensions in the container.` 

Perhaps these docs need to be updated?

Also, it seems this topic would be relevant in the context of [the "Developing inside a container on a remote Docker host" topic](https://code.visualstudio.com/docs/remote/containers-advanced#_developing-inside-a-container-on-a-remote-docker-host).

Should this doc request be a new issue?

**Comment by chrmarti** [comment]
The cache is for uninstalled extensions, they still need to be installed in each container, so the quoted documentation seems to be correct.

Is the permissions issue you observe specific to Docker EE? I'm not familiar with the additional capabilities there, is a volume owned by the user who created it and others cannot reuse it? /cc @bamurtaugh

**Comment by kennethredler** [comment]
Thank you for the feedback, @chrmarti . 😃

Is this "cache for uninstalled extensions" documented somewhere? Perhaps that will help improve understanding.

Yes Docker EE enforces the permission where I may create a volume named `vscode` in a "collection" where I have permission. Another user may see that volume but not have permission to mount it to a container.

Reference [Docker EE Access Control Model](https://cc.bingj.com/cache.aspx?q=docker+EE+access+control+%22collection%22&d=4612545983218021&mkt=en-US&setlang=en-US&w=PJ-RbSnkjPMxixxO417Bvg60RdziX5LP).

**Comment by bamurtaugh** [comment]
Thanks for highlighting the spot where "Cache Volume" could make sense to highlight in docs @kennethredler!

@chrmarti after a quick search through our containers and advanced containers docs, I don't see a current mention of this setting, and as [the section](https://code.visualstudio.com/docs/remote/containers-advanced#_avoiding-extension-reinstalls-on-container-rebuild) @kennethredler linked currently reads, it sounds like "Cache Volume" would make sense to include there. Maybe it makes sense to clarify at the beginning of that section Cache Volume versus creating a named local volume? 

I think I could also use some clarification here (i.e. what does "The cache is for uninstalled extensions, they still need to be installed in each container" exactly mean? The setting description sounds like it's a way to avoid reinstalling VS Code Server and extensions, like the local "named volume" mount does).

**Comment by chrmarti** [comment]
That cache volume comes with its own folder structure and is currently not meant to be directly mounted where the extensions are installed in the container (this is unlike the 'named local volume' in that section of the documentation). It contains the VSIX files of extensions, these then still need to be installed in the container's extensions folder. So it's a cache for avoiding extension downloads. (It also caches server downloads and server installs, we just can't cache extension installs in a reliable way it seems. /cc @sandy081 in case this is changing.)

The 'named local volume' approach in that section of the documentation still makes sense and can be seen as a workaround for the cache volume not supporting caching extension installs.

**Comment by kennethredler** [comment]
Does this still `needsmoreinfo` @chrmarti ?

A workaround was identified.

Will the setting and the capability it controls be documented?

Is the original feature request still valuable/viable?

Thank you!

**Comment by chrmarti** [comment]
@kennethredler Marking as feature request for the setting to control the cache volume name. Thanks.
